toolchain: mwdt: add some compile options to reduce warning

IRISZZW opened the pull request:
When compile zephyr project with arcmwdt toolchain, there are some harmless but noisy compiling warnings,
add some compile options:
`-Wno-typedef-redefinition` 

to reduce warning.

```
warning: redefinition of typedef 'k_tid_t' is a C11 feature [-Wtypedef-redefinition]
```

evgeniy-paltsev commented:
Hmm, any clue about why do we need '-Wno-incompatible-pointer-types'?

evgeniy-paltsev commented:
BTW, your commit doesn't touch '-Wno-incompatible-pointer-types' at all.

IRISZZW commented:
> Hmm, any clue about why do we need '-Wno-incompatible-pointer-types'?

It might be fixed by some code changes in latest commits, I can't reproduce it, so I remove it.

abrodkin reviewed (changes requested):
Now when there's only 1 warning being muted we may explicitly mention in in the commit title.
Also it would be really useful to have an explanation of why that redefinition happens and maybe even insert an extract from real code which requires that treatment.

IRISZZW commented:
> Now when there's only 1 warning being muted we may explicitly mention in in the commit title.
> Also it would be really useful to have an explanation of why that redefinition happens and maybe even insert an extract from real code which requires that treatment.

The redefinition is caused by this PR : [userspace: memory domain cleanups & arch-specific data support](https://github.com/zephyrproject-rtos/zephyr/pull/27687):

```
1 warning generated.
[106/114] Building C object zephyr/kernel/CMakeFiles/kernel.dir/idle.c.obj
In file included from zephyr/kernel/idle.c:7:
../include\kernel.h:648:26: warning: redefinition of typedef 'k_tid_t' is a C11 feature [-Wtypedef-redefinition]
typedef struct k_thread *k_tid_t;
                         ^
../include\app_memory/mem_domain.h:18:26: note: previous definition is here
typedef struct k_thread *k_tid_t;
```
https://github.com/zephyrproject-rtos/zephyr/blob/04fe7c6eeb7880d2a41d1d9d8e17a3d7d427c9ed/include/kernel.h#L648
https://github.com/zephyrproject-rtos/zephyr/blob/04fe7c6eeb7880d2a41d1d9d8e17a3d7d427c9ed/include/app_memory/mem_domain.h#L18

kernel.h include kernel_includes.h:
https://github.com/zephyrproject-rtos/zephyr/blob/04fe7c6eeb7880d2a41d1d9d8e17a3d7d427c9ed/include/kernel.h#L17
kernel_includes.h include mem_domain.h:
https://github.com/zephyrproject-rtos/zephyr/blob/04fe7c6eeb7880d2a41d1d9d8e17a3d7d427c9ed/include/kernel_includes.h#L45

Maybe we need to do some changes in codes, but it's interesting why only mwdt toolchain will treat this as redefinition warning.

tejlmand commented:
@andrewboie could you comment here ^^^

evgeniy-paltsev commented:
Is it possible that it is the same issue?
https://patchwork.kernel.org/patch/10920427/

andrewboie commented:
> @andrewboie could you comment here ^^^

I needed k_tid_t for that header, without pulling in the entire kernel.h to avoid circular deps. GCC was fine with this (probably because the definition is unchanged), but it seems MWDT is not.

Probably the best solution is to create a thread.h header with all the thread definitions and have both kernel.h and mem_domain.h include it

evgeniy-paltsev commented:
@andrewboie, thanks for explanation. It's most likely not MWDT thing but CLANG thing (MWDT is CLANG based) given that link about '-Wno-typedef-redefinition' adding in Linux kernel for CLANG.

@IRISZZW could you please update the commit message? Otherwise ACK.

IRISZZW committed:
toolchain: arcmwdt: add compile options: -Wno-typedef-redefinition

Without the -Wno-typedef-redefinition option, arcmwdt (clang based)
complains if a typedef gets redefined in gnu99  mode (since this is
officially a C11 feature). While new versions of GCC do not seem to
issue this warning in gnu99 mode anymore. So some existing code
with typedef redefined which works well with GCC will issue this
warning. As this warning is not fatal, so simply shut it off for good.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

tejlmand reviewed (approved):
LGTM, but I think it would be nice to create a followup task on this:
> Probably the best solution is to create a thread.h header with all the thread definitions and have both kernel.h and mem_domain.h include it. May itself require some untangling.

so that we can re-enable this warning.
What do you think @andrewboie  ?

carlescufi commented:
@abrodkin could you review this again?

nashif commented:
attempt to fix this proper was submitted here: https://github.com/zephyrproject-rtos/zephyr/pull/29359
